oeps/oep-0017-feature-toggles.rst

```diff
@@ -0,0 +1,927 @@
+requires that every app that uses waffle_utils_ declares its usages in a standard
+module (i.e., config.py) or configure its location (in its apps.py module).
+
+.. _Django App Plugin: https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/plugins/README.rst
+
+Req 12: Report
+^^^^^^^^^^^^^^
+
+In order to provide a useful and informative administrative report of the
+existing feature toggles in the system, the framework needs to be able to
+present the following information for each toggle.
+
+.. list-table::
+   :header-rows: 1
+
+   * - Report data
+     - Purpose
+     - Data source
+   * - Description
+     - Brief human-readable information about its usage and context.
+     - In code, by developer
+   * - Feature Category
+     - Optional field to group interdependent toggles.
+     - In code, by developer
+   * - All Use Cases
+     - Lists one or more `Use Cases`_ to specify all expected usages of this toggle.
+     - In code, by developer
+   * - Current Use Case(s)
+     - A subset of "All Use Cases" to specify the current `Use Cases`_ of this toggle.
+     - In code, by developer; optionally editable via admin interface.
+   * - Toggle Type
+     - One of Switch, Rollout, or Group to further clarify the toggle's usage.
+     - In code, by developer
+   * - Created in Code Date
+     - Required field to specify the date the toggle was added to the codebase; to easily find all stale toggles.
+     - In code, by developer
+   * - Expiration Date
+     - Optional field to specify target date of removal; to easily find all expired toggles.
+     - In code, by developer
+   * - Current Setting(s)
+     - Summary of the current configuration and value of the feature toggle; to easily evaluate its readiness to transition or retire.
+     - Derived from relational tables
+   * - First Modified Time
+     - Date the toggle was first set in the system; to get the starting date of its use.
+     - Derived from relational tables
+   * - Last Modified Time
+     - Date the toggle was last set/unset in the system; to easily find all unused toggles.
+     - Derived from relational tables
+ 
+Testing
+-------
+
+Words of Caution
+~~~~~~~~~~~~~~~~
+
+As `James McKay puts it`_:
+
+  "Visible or not, you are still deploying code into production that you know for 
+  a fact to be buggy, untested, incomplete and quite possibly incompatible with 
+  your live data. Your if statements and configuration settings are themselves 
+  code which is subject to bugs – and furthermore can only be tested in production.
+  ...
+  Your features may not be as isolated from each other as you thought they were, 
+  and you may end up deploying bugs to your production environment.”
+
+.. _James McKay puts it: https://jamesmckay.net/2011/07/why-does-martin-fowler-not-understand-feature-branches/
+
+Testing Best Practices
+~~~~~~~~~~~~~~~~~~~~~~
+
+Given that, here are best practices for testing a Feature Toggle:
```

Comment:
Either in this section and/or the section about waffle_utils, you may want to note that it includes testing utilities.

Reply:
I consider this to be implementation-specific and can go in a good README file within waffle_utils.  Right now, this OEP links to [its __init__ file](https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/waffle_utils/__init__.py), which mentions testing utilities, but should be expanded further.